src/mongo_odm/nvp.hpp

```diff
@@ -74,17 +159,593 @@ struct Nvp {
     constexpr Nvp(T Base::*t, const char* name) : t(t), name(name) {
     }
 
+    /**
+     * Assignment operator that creaptes an update expression to set the field to a given value.
+     */
+    constexpr UpdateExpr<Nvp<Base, T>, no_opt_type> operator=(const no_opt_type& val) const {
+        return {*this, val, "$set"};
+    }
+
+    virtual std::string get_name() const {
+        return name;
+    }
+
     T Base::*t;
     const char* name;
 };
 
+/**
+ * Class that represents a name-value pair for a field of an object that is a member of another
+ * object.
+ * The name of this name-value pair will be "<parent-name>.field_name".
+ * The parenet can be either an Nvp, or another NvpChild.
+ * @tparam Base The type of the object that contains this field
+ * @tparam T    The type of the field
+ * @tparam Parent The type of the parent name-value pair, either an Nvp<...> or NvpChild<...>.
+ */
+template <typename Base, typename T, typename Parent>
+class NvpChild : public NvpCRTP<NvpChild<Base, T, Parent>, T> {
+   public:
+    // In case this field is wrapped in an optional, store the underlying type.
+    using no_opt_type = remove_optional_t<T>;
+    using type = T;
+
+    constexpr NvpChild(T Base::*t, const char* name, Parent parent)
+        : t(t), name(name), parent(parent) {
+    }
+
+    /**
+     * Creates an update expression that sets the field to the given value.
+     */
+    constexpr UpdateExpr<NvpChild<Base, T, Parent>, no_opt_type> operator=(
+        const no_opt_type& val) const {
+        return {*this, val, "$set"};
+    }
+
+    /**
+     * Returns the qualified name of this field in dot notation, i.e. "parent.child".
+     * @return A string containing the name of this field in dot notation.
+     */
+    std::string get_name() const {
```

Comment:
The core builder doesn't really work that way at the moment (it wants the key all at once).

There is one way we could skip all the runtime overhead.  If we changed the Nvp's to include a size_t non-type parameter in their template list, and used that to store the key, we could recursively make bigger name buffers for the NvpChild'ren as well, which would allow us to always have a fully formed key on hand as needed.

We'd find that value by writing a constexpr strlen and using that in the macros we use to tag NVPs.